packages/tailwindcss/src/compat/dark-mode.ts

```diff
@@ -0,0 +1,53 @@
+import type { PluginAPI } from '../plugin-api'
+import type { ResolvedConfig } from './config/types'
+
+export function darkModePlugin({ addVariant, config }: PluginAPI) {
+  let darkMode = config('darkMode', null) as ResolvedConfig['darkMode']
+  let [mode, selector = '.dark'] = Array.isArray(darkMode) ? darkMode : [darkMode]
+
+  if (mode === 'variant') {
+    let formats
+
+    if (Array.isArray(selector)) {
+      formats = selector
+    } else if (typeof selector === 'function') {
+      formats = selector
+    } else if (typeof selector === 'string') {
+      formats = [selector]
+    }
+
+    if (Array.isArray(formats)) {
+      for (let format of formats) {
+        if (format === '.dark') {
+          mode = false
+          console.warn('darkmode-variant-without-selector', [
+            'When using `variant` for `darkMode`, you must provide a selector.',
+            'Example: `darkMode: ["variant", ".your-selector &"]`',
+          ])
```

Comment:
This might need some tweaking 😅

```suggestion
          console.warn(
            'When using `variant` for `darkMode`, you must provide a selector.',
            'Example: `darkMode: ["variant", ".your-selector &"]`',
          )
```

In v3 we use the key to make sure it's only generated once. In this case that's fine because this is only loaded once anyway.
The array was used to pretty print the output with some prefix for each line.